runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -438,34 +467,63 @@ TR_J9ServerVM::cacheField(J9Class *ramClass, int32_t cpIndex, J9Class *declaring
    }
 
-bool
-TR_J9ServerVM::jitStaticsAreSame(TR_ResolvedMethod *method1, I_32 cpIndex1, TR_ResolvedMethod *method2, I_32 cpIndex2)
-   {
-   JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
+void
+TR_J9ServerVM::cacheFields(TR_ResolvedJ9Method *owningMethod)
+   {
+   // 1. Iterate through bytecodes and look for loads/stores
+   // If the corresponding field or static is not cached, add it
+   // to the list of fields that will be sent to the client in one batch.
+   TR_J9ByteCodeIterator bci(0, owningMethod, this, _compInfoPT->getCompilation());
+   std::vector<int32_t> cpIndices;
+   std::vector<uint8_t> isStaticField;
+   J9Class *ramClass = owningMethod->constantPoolHdr();
+   for(TR_J9ByteCode bc = bci.first(); bc != J9BCunknown; bc = bci.next())
+      {
+      J9Class *declaringClass;
+      UDATA field;
+      int32_t cpIndex = bci.next2Bytes();
+      bool isField = false;
+      bool isStatic;
+      if (bc == J9BCgetfield || bc == J9BCputfield)
+         {
+         isField = true;
+         isStatic = false;
+         }
+      else if (bc == J9BCgetstatic || bc == J9BCputstatic)
+         {
+         isField = true;
+         isStatic = true;
+         }
+      if (isField && !getCachedField(ramClass, cpIndex, &declaringClass, &field))
+         {
+         cpIndices.push_back(cpIndex);
+         isStaticField.push_back(isStatic);
+         }
+      }
 
-   // Pass pointers to client mirrors of the ResolvedMethod objects instead of local objects
-   TR_ResolvedJ9JITServerMethod *serverMethod1 = static_cast<TR_ResolvedJ9JITServerMethod*>(method1);
-   TR_ResolvedJ9JITServerMethod *serverMethod2 = static_cast<TR_ResolvedJ9JITServerMethod*>(method2);
-   TR_ResolvedMethod *clientMethod1 = serverMethod1->getRemoteMirror();
-   TR_ResolvedMethod *clientMethod2 = serverMethod2->getRemoteMirror();
+   // If there's just one field, it's faster to get it through regular means,
+   // to avoid overhead of vectors
+   int32_t numFields = cpIndices.size();
+   if (numFields < 2)
+      return;
 
-   bool result = false;
+   // 2. Send a message to get info for all fields
+   JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
+   stream->write(
+      JITServer::MessageType::VM_getFields,
+      static_cast<TR_ResolvedJ9JITServerMethod *>(owningMethod)->getRemoteMirror(),
+      cpIndices,
+      isStaticField);
+   auto recv = stream->read<std::vector<J9Class *>, std::vector<UDATA>>();
 
-   bool sigSame = true;
-   if (serverMethod1->staticsAreSame(cpIndex1, serverMethod2, cpIndex2, sigSame))
+   // 3. Cache all received fields
+   auto &declaringClasses = std::get<0>(recv);
+   auto &fields = std::get<1>(recv);
+   TR_ASSERT(numFields == declaringClasses.size(), "Number of received fields does not match the requested number");
+   for (int32_t i = 0; i < numFields; ++i)
       {
-      result = true;
+      cacheField(ramClass, cpIndices[i], declaringClasses[i], fields[i]);
```

Comment:
Maybe we should enter a critical section for this entire for loop ( OMR::CriticalSection getRemoteROMClass(_compInfoPT->getClientData()->getROMMapMonitor());)
The reason is, acquiring a monitor recursively (while have it) is faster than acquiring it from scratch.

Reply:
Done